# Optimizing User Experience: Behavioral Insights and A/A/B Testing in a Food Delivery App

---
# Table of Contents <a id='back'></a>

[Introduction](#mulai)
1. [Understanding the Sales Funnel](#persiapan)
    * [Data Preparation](#persiapandata)
    * [Data Learning](#pelajaridata)
    * [Funnel Exploration](#pelajarifunnel)
2. [A/A/B Testing Experiment](#eksperimen)
3. [Conclusion & Recommendation](#kesimpulan)

---
# Introduction
<a id="mulai"></a>

In the ever-evolving digital era, understanding user behavior has become crucial for every company, especially for technology-based startups such as the food company featured in this case study. By gaining insight into how users interact with their application, the company can make better decisions in product design, improve user experience, and ultimately increase their business success rate.

In this project, we will discuss user behavior analysis for a startup that sells food products. Our approach consists of several stages:

1. **Understanding the Sales Funnel**: We will examine the sales funnel of this company’s application. The sales funnel will help us understand the user journey from the initial stage to the point of purchase, as well as identify potential drop-off points where users are lost.

2. **A/A/B Testing Experiment**: We will introduce the A/A/B testing experiment conducted by the company to evaluate changes in their app, particularly regarding font design. We will analyze the results of this experiment to determine whether the change had a positive or negative impact on user experience.

3. **Conclusion and Recommendations**: We will summarize our findings from the user behavior analysis and the A/A/B testing results, and provide recommendations for the company on how to optimize their product.

Thus, this project will not only offer deep insights into user behavior and experimental outcomes but also assist the company in making smarter decisions for developing and enhancing their product.

[Back to Table of Contents](#back)

<a id='persiapan'></a>

---
# Understanding the Sales Funnel

## Persiapan Data
<a id="persiapandata"></a>

In [6]:
import pandas as pd

In [7]:
data = pd.read_csv('data/food_app_user_behavior_data.csv')

In [8]:
# show first 5 rows of data
data.head()

EventName\tDeviceIDHash\tEventTimestamp\tExpId
0  MainScreenAppear\t4575588528974610257\t1564029...
1  MainScreenAppear\t7416695313311560658\t1564053...
2  PaymentScreenSuccessful\t3518123091307005509\t...
3  CartScreenAppear\t3518123091307005509\t1564054...
4  PaymentScreenSuccessful\t6217807653094995999\t...

In [9]:
data = pd.read_csv('data/food_app_user_behavior_data.csv')

# change column name
data.columns = ['event_name', 'device_id_hash', 'event_timestamp', 'exp_id']

In [10]:
# general information about data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244126 entries, 0 to 244125
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   event_name       244126 non-null  object
 1   device_id_hash   244126 non-null  int64 
 2   event_timestamp  244126 non-null  int64 
 3   exp_id           244126 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 7.5+ MB


In [11]:
# check the missing values
data.isnull().sum()

event_name         0
device_id_hash     0
event_timestamp    0
exp_id             0
dtype: int64

In [12]:
# convert `event_timestamp` into datetime data type
data['event_date_time'] = pd.to_datetime(data['event_timestamp'], unit='s')

# add date and time columns and a separate column for date
data['date'] = data['event_date_time'].dt.date

# convert `device_id_hash` into string data type
data['device_id_hash'] = data['device_id_hash'].astype(str)

In [13]:
# validate values in `exp_id` column
valid_exp_ids = [246, 247, 248]
invalid_exp_ids = data[~data['exp_id'].isin(valid_exp_ids)]
if not invalid_exp_ids.empty:
    print("Nilai-nilai tidak valid dalam kolom exp_id:", invalid_exp_ids['exp_id'].unique())

data.head()

event_name       device_id_hash  event_timestamp  exp_id  \
0         MainScreenAppear  4575588528974610257       1564029816     246   
1         MainScreenAppear  7416695313311560658       1564053102     246   
2  PaymentScreenSuccessful  3518123091307005509       1564054127     248   
3         CartScreenAppear  3518123091307005509       1564054127     248   
4  PaymentScreenSuccessful  6217807653094995999       1564055322     248   

      event_date_time        date  
0 2019-07-25 04:43:36  2019-07-25  
1 2019-07-25 11:11:42  2019-07-25  
2 2019-07-25 11:28:47  2019-07-25  
3 2019-07-25 11:28:47  2019-07-25  
4 2019-07-25 11:48:42  2019-07-25

Data Preparation Steps:

1. The dataset was confirmed to contain no missing values, ensuring data completeness for subsequent analysis.
2. The `event_timestamp` field was converted to a datetime format to enable accurate and efficient time-based analyses.
3. Additional columns were derived to separately capture the full timestamp, the date, and the time components, facilitating granular temporal analysis.
4. The `device_id_hash` field was cast to a string type. As this identifier is not intended for numerical computation, treating it as a string minimizes the risk of errors during data manipulation and ensures semantic clarity.
5. The values within the `exp_id` column were examined to confirm their validity. Specifically, we verified that only the expected identifiers (246, 247, and 248) were present, and that no extraneous or anomalous values existed in the dataset.alues.

[Back to Table of Contents](#back)

---
## Data Learning
<a id="pelajaridata"></a>

In [14]:
# how many events recorded in log?
total_events = data.shape[0]
print("Total events in log:", total_events)

Total events in log: 244126


In [15]:
# how many users recorded in log?
total_users = data['device_id_hash'].nunique()
print("Total users in log:", total_users)

Total users in log: 7551


In [16]:
# what is the average number of events per user?
avg_events_per_user = total_events / total_users
print("Average events per users:", avg_events_per_user)

Average events per users: 32.33028737915508


In [17]:
# what time period is covered by the data?
min_date = data['event_date_time'].min()
max_date = data['event_date_time'].max()
print("Minimum date:", min_date)
print("Maximum date:", max_date)

Minimum date: 2019-07-25 04:43:36
Maximum date: 2019-08-07 21:15:17


In [18]:
# histogram based on date and time
mport matplotlib.pyplot as plt

plt.hist(data['event_date_time'], bins=30, color='orange', edgecolor='black')
plt.xlabel('Date and Time')
plt.ylabel('Frequency')
plt.title('Histogram Date and Time Event')
plt.xticks(rotation=45)
plt.show()

SyntaxError: invalid syntax (1937635139.py, line 2)

In [ ]:
# counting total events per date
events_per_date = data.groupby('date').size()

# visualize event distribution per date
plt.figure(figsize=(10, 6))
plt.plot(events_per_date.index, events_per_date.values, marker='o', linestyle='-')
plt.xlabel('Date')
plt.ylabel('Total Events')
plt.title('Event Distribution per Date')
plt.xticks(rotation=45)
plt.grid(True)
plt.show()

In [ ]:
# ensure that we have users from three experiment group
exp_group = data['exp_id'].unique()
print("Experiment group recorded in data:", exp_group)

In [ ]:
data.head()

1. Total number of events in the log: 244,126
2. Total number of users recorded: 7,551
3. Average number of events per user: 32.33
4. Earliest timestamp in the dataset: July 25, 2019, 04:43:36
5. Latest timestamp in the dataset: August 7, 2019, 21:15:17
6. Experimental groups represented in the data: [246, 248, 247]

[Back to Table of Contents](#back)

---
## Funnel Exploration
<a id="pelajarifunnel"></a>

In [ ]:
# identify the events present in the log and their frequencies of occurrence.
event_counts = data['event_name'].value_counts()
print("Event frequencies of occurrence:")

event_counts

In [ ]:
# The number of users who performed each action
user_counts = data.groupby('event_name')['device_id_hash'].nunique().sort_values(ascending=False)
print("The number of users who performed each action:")

user_counts

In [ ]:
# The number of users who performed each action
user_per_event = data.groupby('event_name')['device_id_hash'].nunique().reset_index()
user_per_event.columns = ['event_name', 'unique_users']
user_per_event = user_per_event.sort_values(by='unique_users', ascending=False)

In [ ]:
# user proportion who perform one time
user_total = data['device_id_hash'].nunique()
one_time_user = user_per_event['unique_users'].sum()
prop_one_time_user = one_time_user / user_total * 100
print("User proportion who perform at least one time:", prop_one_time_user)

In [ ]:
# event funnel: the percentage of users who progressed from one stage to the next.
funnel = user_per_event.copy()
funnel['retention_rate'] = funnel['unique_users'].pct_change() * 100
funnel = funnel.dropna()

In [ ]:
# identify the stage at which user drop-off occurs
loss = funnel[funnel['retention_rate'] < 0]['event_name'].iloc[0]
print("Tahap di mana kehilangan pengguna terjadi:", loss)

In [ ]:
# the percentage of users who successfully completed all stages
first_user = funnel.iloc[0]['unique_users']
last_user = funnel.iloc[-1]['unique_users']
persentase_penyelesaian = (last_user / first_user) * 100
print("The percentage of users who successfully completed all stages:", persentase_penyelesaian)

1. Event Frequency:
- MainScreenAppear: 119,205 occurrences
- OffersScreenAppear: 46,825 occurrences
- CartScreenAppear: 42,731 occurrences
- PaymentScreenSuccessful: 34,313 occurrences
- Tutorial: 1,052 occurrences

2. Number of Users who performed each action:
- MainScreenAppear: 7,439 users
- OffersScreenAppear: 4,613 users
- CartScreenAppear: 3,749 users
- PaymentScreenSuccessful: 3,547 users
- Tutorial: 847 users

3. Proportion of users who performed at least one action: 267.45%
4. Stage where the highest user drop-off occurs: OffersScreenAppear
5. Percentage of users who successfully completed all stages: 18.36%

[Back to Table of Contents](#back)

---
# A/A/B Testing Experiment
<a id="eksperimen"></a>

In [ ]:
# how many users in each group?
group_counts = data.groupby('exp_id')['device_id_hash'].nunique().reset_index()
group_counts.columns = ['exp_id', 'user_counts']
print("Total users in every experiment group:")
print(group_counts)

In [ ]:
# find significant difference between control group
control_group_A = data[(data['exp_id'] == 246) | (data['exp_id'] == 247)]
control_group_B = data[data['exp_id'] == 248]

In [ ]:
# most popular event
most_popular_event = data['event_name'].value_counts().idxmax()

print("Most popular event:", most_popular_event)

In [ ]:
from scipy.stats import chi2_contingency

def analyze_event(df, event_name):
    # Find the number of users who performed each event within each control group.
    control_group_A = data[(data['exp_id'] == 246) | (data['exp_id'] == 247)]
    control_group_B = data[data['exp_id'] == 248]

    users_control_A = control_group_A[control_group_A['event_name'] == event_name]['device_id_hash'].nunique()
    users_control_B = control_group_B[control_group_B['event_name'] == event_name]['device_id_hash'].nunique()

    # Count the percentage
    group_counts = data.groupby('exp_id')['device_id_hash'].nunique().reset_index()
    group_counts.columns = ['exp_id', 'user_counts']
    percent_control_A = (users_control_A / group_counts.loc[group_counts['exp_id'] == 246, 'user_counts'].values[0]) * 100
    percent_control_B = (users_control_B / group_counts.loc[group_counts['exp_id'] == 248, 'user_counts'].values[0]) * 100

    # Statistic test
    contingency_table = pd.crosstab(df['exp_id'], df['event_name'])
    chi2, p, _, _ = chi2_contingency(contingency_table)

    # Show the outcome
    print(f"Peristiwa: {event_name}")
    print(f"Total users in control group A who performed the event: {users_control_A}")
    print(f"Total users in control group B who performed the event: {users_control_B}")
    print(f"Percentage of users in control group A: {percent_control_A:.2f}%")
    print(f"Percentage of users in control group B: {percent_control_B:.2f}%")
    print(f"Statistic test chi-squared: chi2 = {chi2}, p-value = {p}")
    if p < 0.05:
        print("The difference between the groups is statistically significant.")
    else:
        print("The difference between the groups is not statistically significant.")
    print("\n")

# find the most popular event
most_popular_event_control = control_group_A['event_name'].value_counts().idxmax()
most_popular_event_experimental = control_group_B['event_name'].value_counts().idxmax()

# analyze and test the statistic for each event
for event in data['event_name'].unique():
    analyze_event(data, event)

In [ ]:
# count the number of statistic hypotheses test
total_tests = 2 * len(data['event_name'].unique())
print("Total statistic hypotheses test performed:", total_tests)

[Back to Table of Contents](#back)

---
# Conclusion & Recommendation
<a id="kesimpulan"></a>

## Conclusion

### Data Preparation Steps
1. **Reading the Data File**: The CSV data file was read using `pd.read_csv()`.
2. **Column Name Adjustment**: Column names were modified as needed using the `.columns` attribute.
3. **General Data Overview**: A general overview of the dataset was examined using `.info()` to ensure no missing values and to review column data types.
4. **Missing Value Check**: Missing values were checked using `.isnull().sum()` to confirm data completeness.
5. **Data Type Conversion**: The `event_timestamp` column was converted to datetime format using `pd.to_datetime()`. Additionally, the `device_id_hash` column was converted to a string, as device IDs do not require mathematical operations.
6. **Date Column Addition**: A new column `event_date_time` was added to store timestamps in datetime format. Another column, date, was created to extract and store only the date component.
7. **Validation of exp_id Values**: A check was conducted to ensure that the `exp_id` column contained only valid values consistent with the expected experimental groups.
8. **Displaying the Data**: The first five rows of the prepared dataset were displayed using `.head()` to verify that all changes had been applied correctly.

Following these data preparation steps, the dataset was ready for further analysis. No missing values were found, all columns were properly typed, and the inclusion of date fields and validation of experimental groups ensured high data quality. The dataset was thus well-suited for subsequent analysis and modeling.

### Data Analysis
1. **Total Events in the Log**: The log contains a total of 244,126 events.
2. **Total Users in the Log**: There are 7,551 distinct users recorded in the log.
3. **Average Events per User**: On average, each user generated approximately 32.33 events.
4. **Data Time Range**: The dataset covers the period from July 25, 2019, to August 7, 2019.
5. **Histogram of Event Timestamps**: A histogram visualizes the temporal distribution of events, revealing fluctuations in event frequency over time.
6. **Event Distribution by Date**: A separate plot illustrates the daily distribution of events, highlighting spikes in activity that may correspond to specific events or changes in the app.
7. **Experiment Groups**: Users are represented across all three experimental groups—246, 247, and 248—ensuring balanced representation for analysis.

**Conclusion:**
- The dataset contains a substantial volume of events over approximately two weeks.
- The average number of events per user is around 32.
- Daily variations in activity suggest user behavior may be influenced by external or app-related factors.
- All experimental groups are present in the data, allowing for valid comparative analysis.

### Funnel Analysis
From the funnel analysis, the following conclusions can be drawn:

1. **Event Frequency**: MainScreenAppear is the most frequently occurring event, followed by OffersScreenAppear, CartScreenAppear, PaymentScreenSuccessful, and finally Tutorial.
2. **Number of Users per Event**: The number of users performing each event decreases through the funnel, with the highest count at MainScreenAppear and the lowest at Tutorial.
3. **Proportion of Users Performing at Least One Action**: Approximately 267.45%, which suggests that many users performed multiple actions.
4. **Event Funnel**: The percentage of users progressing from one stage to the next drops notably at the OffersScreenAppear stage, indicating significant user attrition.
5. **User Drop-off**: The most substantial drop-off occurs at the OffersScreenAppear stage. This may point to a usability or engagement issue at this point in the app.
6. **Users Completing the Entire Funnel**: Around 18.36% of users successfully completed all funnel stages, indicating a challenge in retaining users through to conversion.

### A/A/B Testing
Based on the A/A/B testing analysis, the following insights and recommendations emerge:

1. User Distribution Across Groups: The three experimental groups are relatively balanced, each with approximately 2,500 users.
2. Most Popular Event: The MainScreenAppear event is the most frequently recorded across all groups.
3. Event Analysis: All key events showed statistically significant differences between the control and experimental groups. This is supported by very low p-values (< 0.05) from chi-squared tests, indicating that group differences are unlikely due to chance.
4. Total Hypothesis Tests Conducted: Ten hypothesis tests were performed—covering each event across the control (A), second control (A), and experimental (B) groups, as per standard A/A/B testing protocols.

## Recommendations
1. **Validate Findings**: Despite statistically significant results, it's important to validate these findings by considering other variables that may influence user behavior, such as demographics or historical activity.
2. **Deeper Analysis**: Conduct further analysis to explore factors behind the observed group differences. This may include segmenting users by prior behavior or external conditions.
3. **Experiment Optimization**: If the tested changes led to positive outcomes (e.g., higher conversion or retention rates), consider rolling out those changes more broadly.
4. **Iterate and Monitor**: A/A/B testing is inherently iterative. After implementing changes, continue monitoring user behavior and perform follow-up tests to refine the product further.
5. **Control for Multiple Testing**: Given the number of hypothesis tests, be cautious of Type I errors. Employing a conservative significance threshold (e.g., 0.05 or lower) helps mitigate the risk of false positives due to multiple comparisons.

[Back to Table of Contents](#back)